warp/yul/AstTools.py

```diff
@@ -0,0 +1,420 @@
+        self.pos += 1
+        body = self.parse_block()
+
+        return ast.FunctionDefinition(
+            name=fun_name, parameters=params, return_variables=returns, body=body
+        )
+
+    def parse_if(self) -> ast.If:
+        tabs = self.get_tabs()
+        assert self.get_word(tabs) == "If:", "This node should be of type If"
+        self.pos += 1
+
+        condition = self.parse_expression()
+        body = self.parse_block()
+        else_body = None
+        if self.get_tabs() > tabs:
+            else_body = self.parse_block()
+
+        return ast.If(condition=condition, body=body, else_body=else_body)
+
+    def parse_case(self) -> ast.Case:
+        tabs = self.get_tabs()
+        assert self.get_word(tabs) == "Case:", "This node should be of type Case"
+        self.pos += 1
+
+        try:
+            value = self.parse_literal()
+        except AssertionError:
+            assert (
+                self.get_tabs() == tabs + 1 and self.get_word(tabs + 1) == "Default"
+            ), "The value must be a literal or None (when it's the default case)"
+            value = None
+            self.pos += 1
+
+        return ast.Case(value=value, body=self.parse_block())
+
+    def parse_switch(self) -> ast.Switch:
+        tabs = self.get_tabs()
+        assert self.get_word(tabs) == "Switch:", "This node should be of type Switch"
+        self.pos += 1
+
+        return ast.Switch(
+            expression=self.parse_expression(),
+            cases=self.parse_list(tabs, self.parse_case),
+        )
+
+    def parse_for_loop(self) -> ast.ForLoop:
+        tabs = self.get_tabs()
+        assert self.get_word(tabs) == "ForLoop:", "This node should be of type ForLoop"
+        self.pos += 1
+
+        return ast.ForLoop(
+            pre=self.parse_block(),
+            condition=self.parse_expression(),
+            post=self.parse_block(),
+            body=self.parse_block(),
+        )
+
+    def parse_break(self) -> ast.Break:
+        tabs = self.get_tabs()
+        assert self.get_word(tabs) == "Break", "This node should be of type Break"
+        self.pos += 1
+
+        return ast.Break()
+
+    def parse_continue(self) -> ast.Continue:
+        tabs = self.get_tabs()
+        assert self.get_word(tabs) == "Continue", "This node should be of type Continue"
+        self.pos += 1
+
+        return ast.Continue()
+
+    def parse_leave(self) -> ast.Leave:
+        tabs = self.get_tabs()
+        assert self.get_word(tabs) == "Leave", "This node should be of type Leave"
+        self.pos += 1
+
+        return ast.LEAVE, pos
+
+    def parse_node(self) -> ast.Node:
+        tabs = self.get_tabs()
+        node_type_name = self.get_word(tabs).split(":")[0]
+        parser_name = f"parse_{self.get_name(node_type_name)}"
+
+        try:
+            return getattr(self, parser_name)()
+        except:
+            raise WarpException("Wrong node type name!")
```

Comment:
When I did this with AstVisitor, I got a lot of nasty false positives. I suggest doing the following
```suggestion
        parser = getattr(self, parser_name)
        if not parser: raise WarpException(bla-bla)
        return parser()
        
        try:
            return getattr(self, parser_name)()
        except:
            raise WarpException("Wrong node type name!")
```